sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py

```diff
@@ -894,7 +897,10 @@ def process_bundle(self,
       finalize_request = beam_fn_api_pb2.InstructionRequest(
           finalize_bundle=beam_fn_api_pb2.FinalizeBundleRequest(
               instruction_id=process_bundle_id))
-      self._worker_handler.control_conn.push(finalize_request)
+      finalize_response = self._worker_handler.control_conn.push(
+          finalize_request).get()
```

Comment:
this .get() was necessary to prevent a race condition where the pipeline could shutdown before finalization happened.